diff --git a/src/freedreno/ir3/ir3_sched.c b/src/freedreno/ir3/ir3_sched.c
--- a/src/freedreno/ir3/ir3_sched.c
+++ b/src/freedreno/ir3/ir3_sched.c
@@ -500,6 +500,8 @@
 
       for (unsigned i = 0; i < ir->baryfs_count; i++) {
          struct ir3_instruction *baryf = ir->baryfs[i];
+         if (baryf->block != instr->block)
+            continue;
          if (baryf->flags & IR3_INSTR_UNUSED)
             continue;
          if (!is_scheduled(baryf)) {
